### PR TITLE
PWN-5812 - Integrate History Service

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-5812

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/history/repository/remote/MoonpayHistoryRemoteRepository.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    override suspend fun loadHistory(limit: Int, offset: Int): List<HistoryTransaction> {
+        if (sellEnabledFeatureToggle.isFeatureEnabled) {
+            val newTransactions = repository.getUserSellTransactions(tokenKeyProvider.publicKey.toBase58Instance())
+            if (!allTransactions.containsAll(newTransactions)) {
+                allTransactions.addAll(newTransactions)
+            }
+            return allTransactions
+        }
+        return emptyList()
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/MoonpayHistoryRemoteRepository.kt:26]: Can you check first if `isFeatureEnabled`?

```
if (!sellEnabledFeatureToggle.isFeatureEnabled) return emptyList()

val newTransactions = repository.getUserSellTransactions(tokenKeyProvider.publicKey.toBase58Instance())
if (!allTransactions.containsAll(newTransactions)) {
     allTransactions.addAll(newTransactions)
}
return allTransactions
```

File: app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRemoteRepository.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+private const val REQUEST_PARAMS_OFFSET = "offset"
+private const val REQUEST_PARAMS_SIGNATURE = "signature"
+private const val REQUEST_PARAMS_NAME = "get_transactions"
+
+class RpcHistoryRemoteRepository(
+    private val historyApi: RpcHistoryServiceApi,
+    private val tokenKeyProvider: TokenKeyProvider,
+    private val historyServiceSignatureFieldGenerator: HistoryServiceSignatureFieldGenerator,
+    private val converter: RpcHistoryTransactionConverter
+) : HistoryRemoteRepository {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/RpcHistoryRemoteRepository.kt:23]: `HistoryRepository`

File: app/src/main/java/org/p2p/wallet/history/ui/detailsbottomsheet/HistoryTransactionDetailsBottomSheetPresenter.kt [hunk 3]
```diff
@@ -28,33 +26,33 @@ class HistoryTransactionDetailsBottomSheetPresenter(
...
         launch {
             view?.showLoading(isLoading = true)
-            when (state) {
-                is TransactionDetailsLaunchState.History -> loadDetailsByTransaction(state.transaction)
-                is TransactionDetailsLaunchState.Id -> loadDetailsById(state)
-            }
+//            when (state) {
+//                is TransactionDetailsLaunchState.History -> loadDetailsByTransaction(state.transaction)
+//                is TransactionDetailsLaunchState.Id -> loadDetailsById(state)
+//            }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/detailsbottomsheet/HistoryTransactionDetailsBottomSheetPresenter.kt:32]: ?

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt [hunk 4]
```diff
@@ -49,53 +33,36 @@ class HistoryListViewPresenter(
...
-            view?.scrollToTop()
+    override fun loadHistory() {
+        launch {
+            try {
+                val items = historyInteractor.loadHistory(20, 0)
+                val adapterItems = historyItemMapper.toAdapterItem(items)
+                view?.showHistory(adapterItems)
+                view?.showPagingState(PagingState.Idle)
+            } catch (e: Throwable) {
+                Timber.tag("______").e(e)
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt:46]: forgot to fix it

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt [hunk 5]
```diff
@@ -49,53 +33,36 @@ class HistoryListViewPresenter(
...
 
-        refreshJob = launch {
-            view?.showRefreshing(isRefreshing = true)
-            fetchHistory(isRefresh = true)
-            view?.showRefreshing(isRefreshing = false)
-            view?.scrollToTop()
+    override fun loadHistory() {
+        launch {
+            try {
+                val items = historyInteractor.loadHistory(20, 0)
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt:41]: magic numbers, named params

File: app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt [hunk 6]
```diff
@@ -16,11 +18,24 @@ import org.p2p.wallet.utils.getStatusIcon
...
+    fun toAdapterItem(transactions: List<HistoryTransaction>): List<HistoryItem> {
+        val adapterItems = mutableListOf<HistoryItem>()
+        transactions.forEachIndexed { index, item ->
+            when {
+                item is RpcHistoryTransaction -> {
+                    parse(item, adapterItems)
+                }
+
+                item is SellTransaction -> {
+                }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/sell/interactor/HistoryItemMapper.kt:30]: add todo

**Review by 1sakov — APPROVED** [review]
overall, looks very nice, very simple
well done!

File: app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt [hunk 7]
```diff
@@ -1,13 +1,16 @@
...
 import org.p2p.wallet.moonpay.repository.sell.SellTransactionFiatCurrency
 import org.p2p.wallet.moonpay.serversideapi.response.SellTransactionFailureReason
 import org.p2p.wallet.moonpay.serversideapi.response.SellTransactionStatus
 import org.p2p.wallet.utils.Base58String
+import org.threeten.bp.ZonedDateTime
 
 sealed class SellTransaction(
     val status: SellTransactionStatus
-) {
+) : MoonPayTransaction() {
```
[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt:13]: I don't like inheritance in this case

Better do make this class as 
```
MoonpayHistoryTransaction(
val sellTransaction: SellTransaction
) : HistoryTransaction()
```

File: app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt [hunk 8]
```diff
@@ -144,11 +144,13 @@ class TokenHistoryFragment :
...
-                HistoryTransactionDetailsBottomSheetFragment.show(parentFragmentManager, state)
+            is RpcHistoryTransaction.Swap,
+            is RpcHistoryTransaction.Transfer,
+            is RpcHistoryTransaction.BurnOrMint -> {
+                HistoryTransactionDetailsBottomSheetFragment.show(
+                    fragmentManager = parentFragmentManager,
+                    signature = transaction.getHistoryTransactionId()
+                )
             }
             else -> Timber.e("Unsupported transactionType: $transaction")
```
[the remark this answers is not in the document]
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/history/ui/token/TokenHistoryFragment.kt:155]: I just update imports here, I will be back here in next tasks of history

File: app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt:13]: let's discuss, but in next Pull Requests ))

[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/moonpay/model/SellTransaction.kt:13]: we won't remember this, but ok

File: app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt [hunk 9]
```diff
@@ -49,53 +33,36 @@ class HistoryListViewPresenter(
...
         when (historyItem) {
             is HistoryItem.TransactionItem -> {
-                val historyTransaction = blockChainTransactionsList.content
-                    .firstOrNull { it.signature == historyItem.signature }
-                if (historyTransaction != null) {
-                    view?.onTransactionClicked(historyTransaction)
-                } else {
-                    Timber.e("Transaction not founded for history item! $historyItem")
-                }
+                val item = historyInteractor.findTransactionBySignature(historyItem.signature) ?: error(
```
[the remark this answers is not in the document]
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/history/ui/historylist/HistoryListViewPresenter.kt]: I will change to 'return', but it should not crash

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**